### PR TITLE
Created nextjs app for issue_hub

### LINKED ISSUE #78: write the endpoints for user to login and signup
change the existing to c.json files

### LINKED ISSUE #79: login page and signup page using react
- [x] sigining up an user will redirect u to login page and 
- [x] login in user will redirect u to user page .

### PR DESCRIPTION
**Pull request by sridhar02**
Fixes #78 
Fixes #79

### CODE REVIEW

File: issue.go [hunk 1]
```diff
@@ -23,40 +23,41 @@ type Issue struct {
 	UpdatedAt   time.Time `json:"updated_at,omitempty"`
 }
 
-func GetIssue(db *sql.DB, id int) (Issue, error) {
+func GetIssue(db *sql.DB, userId string) (Issue, error) {
```
Comment (vramana) [inline issue.go]: This function is wrong. GetIssue represent fetching one issue by it's id. You can't fetch one issue by userId. it does not make sense.

Comment (sridhar02) [reply issue.go]: yes i have to change just know identified that mistakee

File: issue.go [hunk 2]
```diff
@@ -46,6 +48,20 @@ func GetIssue(db *sql.DB, Id int) (Issue, error) {
 		fmt.Println(err)
 		return Issue{}, err
 	}
+	var name, username, email, image string
+	row = db.QueryRow(`SELECT name,username,email,image FROM users where id=$1`, userId)
```
Comment (vramana) [inline issue.go]: Use GetUser instead.

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js [hunk 3]
```diff
@@ -74,6 +74,12 @@ const Comment = withStyles(commentStyles)(_Comment);
 const STATUS_OPEN = "Open";
 const STATUS_CLOSED = "Closed";
 
+const LOCK_OPEN = "Unlocked";
```
Comment (vramana) [inline issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: Why is this LOCK_OPEN and LOCK_CLOSE ?

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js [hunk 4]
```diff
@@ -74,11 +74,11 @@ const Comment = withStyles(commentStyles)(_Comment);
 const STATUS_OPEN = "Open";
 const STATUS_CLOSED = "Closed";
 
-const LOCK_OPEN = "Unlocked";
-const LOCK_CLOSE = "Locked";
+const LOCK_LOCK = "Unlocked";
```
Comment (vramana) [inline issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: Incorrect. This is LOCK_UNLOCK

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js [hunk 5]
```diff
@@ -74,11 +74,11 @@ const Comment = withStyles(commentStyles)(_Comment);
...
+const LOCK_UNLOCK = "Locked";
 
-const PIN = "Pinned";
-const UNPIN = "Unpinned";
+const PIN_PIN = "Pinned";
+const UNPIN_UNPIN = "Unpinned";
```
Comment (vramana) [inline issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: PIN_UNPIN. Field name does not change.

Comment (sridhar02) [reply issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: corrected

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js (hunk 4 above)
Comment (sridhar02) [reply issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: corrected

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js (hunk 3 above)
Comment (sridhar02) [reply issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: changed

File: issue.go (hunk 2 above)
Comment (sridhar02) [reply issue.go]: changed

File: issue.go (hunk 1 above)
Comment (sridhar02) [reply issue.go]: changed

File: comment.go [hunk 6]
```diff
@@ -113,19 +112,47 @@ func getCommentHandler(c *gin.Context, db *sql.DB) {
...
+	var issueId int
+	row := db.QueryRow(`select repos.id FROM repos JOIN users ON repos.user_id= users.id 
+         				   WHERE repos.name= $1 AND users.username= $2`, repoName, username)
+	err = row.Scan(&repoId)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	row = db.QueryRow(`select id FROM issues WHERE repo_id= $1 AND issue_number= $2`, repoId, issueNumber)
```
Comment (vramana) [inline comment.go]: WITH

File: issue.go [hunk 7]
```diff
@@ -19,44 +20,54 @@ type Issue struct {
...
 
-	var title, userId, body, repoId, status, createdAt, updatedAt, pinned string
-	var issueNumber int
+	var title, body, repoId, status, createdAt, userId, updatedAt, pinned, lock string
+	var id, issueNumber int
 
-	row := db.QueryRow("SELECT title, user_id,body,repo_id,issue_number,status,pinned,created_at, updated_at FROM issues WHERE id=$1", id)
-	err := row.Scan(&title, &userId, &body, &repoId, &issueNumber, &status, &createdAt, &updatedAt, &pinned)
+	row := db.QueryRow(`SELECT id,title,body,repo_id,issue_number,status,pinned,lock,user_id,created_at, 
+						updated_at FROM issues WHERE id=$1`, Id)
```
Comment (vramana) [inline issue.go]: `id` and `Id` are the same

File: issue.go [hunk 8]
```diff
@@ -19,44 +20,54 @@ type Issue struct {
...
+	err := row.Scan(&id, &title, &body, &repoId, &issueNumber, &status, &pinned, &lock, &userId, &createdAt, &updatedAt)
```
Comment (vramana) [inline issue.go]: id is unnecessary in Scan we already have id as argument

File: issue.go [hunk 9]
```diff
@@ -123,53 +134,92 @@ func DeleteIssue(db *sql.DB, id int) error {
...
+	}
+	if issue.Title != "" {
+		query = query.Set("title", issue.Title)
+	}
+	if issue.Body != "" {
+		query = query.Set("body", issue.Body)
+	}
+	if issue.Pinned != "" {
+		query = query.Set("pinned", issue.Pinned)
+	}
```
Comment (vramana) [inline issue.go:232]: pinned should not be done here in PUT.

Have POST /pin  and DELETE /pin. 

Check Condition POST: When you pin issue, you have to check there are no more than 2 pinned issues already.

File: issue.go [hunk 10]
```diff
@@ -207,3 +257,85 @@ func deleteIssueHandler(c *gin.Context, db *sql.DB) {
 	c.Status(http.StatusNoContent)
 
 }
+
+func putLocKHandler(c *gin.Context, db *sql.DB) {
```
Comment (vramana) [inline issue.go]: putLockHandler

File: issue.go [hunk 11]
```diff
@@ -207,3 +257,85 @@ func deleteIssueHandler(c *gin.Context, db *sql.DB) {
...
+		return
+	}
+	username := c.Param("owner")
+	repoName := c.Param("repo")
+	issueNumber := c.Param("issue_number")
+
+	var Id int
+	row := db.QueryRow(`WITH repo_cte AS (select repos.id FROM repos JOIN users ON repos.user_id= users.id 
+                		WHERE repos.name= $1 AND users.username= $2) select id from issues where repo_id 
+                		in (select id from repo_cte) and issue_number=$3`, repoName, username, issueNumber)
```
Comment (vramana) [inline issue.go]:
```
	row := db.QueryRow(`WITH repo_cte AS (
                        SELECT repos.id FROM repos JOIN users ON repos.user_id= users.id 
                		    WHERE repos.name= $1 AND users.username= $2
                      ) 
                      SELECT id FROM issues 
                      WHERE repo_id in (select id from repo_cte) and issue_number=$3`, 
                      repoName, 
                      username, 
                      issueNumber)
```

File: issue.go [hunk 12]
```diff
@@ -207,3 +257,85 @@ func deleteIssueHandler(c *gin.Context, db *sql.DB) {
...
+
+	issue := Issue{}
+	err = c.BindJSON(&issue)
+	if err != nil {
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	_, err = db.Exec(`UPDATE issues SET lock = $1 WHERE id = $2`, issue.Lock, Id)
```
Comment (vramana) [inline issue.go]: Check if the above CTE can be merged into this.

File: issue.go [hunk 13]
```diff
@@ -207,3 +257,85 @@ func deleteIssueHandler(c *gin.Context, db *sql.DB) {
...
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	c.Status(http.StatusNoContent)
+
+}
+
+func deleteLocKHandler(c *gin.Context, db *sql.DB) {
```
Comment (vramana) [inline issue.go]: deleteLockHandler

File: issue.go [hunk 14]
```diff
@@ -207,3 +257,85 @@ func deleteIssueHandler(c *gin.Context, db *sql.DB) {
...
+	}
+	username := c.Param("owner")
+	repoName := c.Param("repo")
+	issueNumber := c.Param("issue_number")
+
+	var Id int
+	row := db.QueryRow(`WITH repo_cte AS (select repos.id FROM repos JOIN users ON repos.user_id= users.id 
+                		WHERE repos.name= $1 AND users.username= $2) select id from issues where repo_id 
+                		in (select id from repo_cte) and issue_number=$3`, repoName, username, issueNumber)
+	err = row.Scan(&Id)
```
Comment (vramana) [inline issue.go]: Same as above.

File: issue.go [hunk 15]
```diff
@@ -207,3 +257,85 @@ func deleteIssueHandler(c *gin.Context, db *sql.DB) {
...
+	issue := Issue{}
+	err = c.BindJSON(&issue)
+	if err != nil {
+		fmt.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	_, err = db.Exec(`UPDATE issues SET lock = $1 WHERE id = $2`, issue.Lock, Id)
+
```
Comment (vramana) [inline issue.go]: Same as above

File: issue_hub/components/nav.js [hunk 16]
```diff
@@ -0,0 +1,56 @@
+import React from 'react'
```
Comment (vramana) [inline issue_hub/components/nav.js]: Remove this file

File: issue_hub/utils/utils.js [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+                <div className="d-none d-md-block">Issues</div>
+                <div className="d-none d-md-block">Marketplace</div>
+                <div className="d-none d-md-block">Explore</div>
+                <div className={cx(classes.space, "d-none d-md-block")}></div>
+                <NotificationsIcon />
+                <div className="d-none d-md-block">
+                  <img src={user.image} className={classes.userImage} />
+                  <button type="button" className={classes.dropDown}>
+                    <ArrowDropDownIcon onClick="" />
+                  </button>
```
Comment (vramana) [inline issue_hub/utils/utils.js]:
```
                  {user && <button type="button" className={classes.dropDown}>
                    <img src={user.image} className={classes.userImage} />
                    <ArrowDropDownIcon onClick="" />
                  </button>}
```

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js [hunk 18]
```diff
@@ -0,0 +1,574 @@
...
+          issue.status === STATUS_OPEN
+            ? classes.commentOpen
+            : classes.commentClose
+        }
+      >
+        {issue.status === STATUS_OPEN ? (
+          <ErrorOutlineIcon className={classes.statusCloseIcon} />
+        ) : (
+          ""
+        )}
```
Comment (vramana) [inline issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]:
```
{issue.status === STATUS_OPEN &&  <ErrorOutlineIcon className={classes.statusCloseIcon} />}
```

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js [hunk 19]
```diff
@@ -0,0 +1,574 @@
...
+    padding: theme.spacing(1)
+  },
+  username: {
+    fontWeight: "bold",
+    marginRight: theme.spacing(1)
+  }
+});
+
+class _Comment extends Component {
+  render() {
```
Comment (vramana) [inline issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: Functional component

**Review by vramana — CHANGES_REQUESTED** [review]

File: comment.go (hunk 6 above)
Comment (sridhar02) [reply comment.go]: DONE

File: issue.go (hunk 7 above)
Comment (sridhar02) [reply issue.go]: corrected

File: issue.go (hunk 10 above)
Comment (sridhar02) [reply issue.go]: done

File: issue.go (hunk 11 above)
Comment (sridhar02) [reply issue.go]: done

File: issue.go (hunk 14 above)
Comment (sridhar02) [reply issue.go]: done

File: issue.go (hunk 15 above)
Comment (sridhar02) [reply issue.go]: done

File: issue_hub/components/nav.js (hunk 16 above)
Comment (sridhar02) [reply issue_hub/components/nav.js]: done

File: issue_hub/utils/utils.js (hunk 17 above)
Comment (sridhar02) [reply issue_hub/utils/utils.js]: done

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js (hunk 18 above)
Comment (sridhar02) [reply issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: done

File: issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js (hunk 19 above)
Comment (sridhar02) [reply issue_hub/pages/user/[username]/repos/[repo]/issues/[issueNumber].js]: done

File: issue.go (hunk 13 above)
Comment (sridhar02) [reply issue.go]: done

File: issue.go (hunk 12 above)
Comment (sridhar02) [reply issue.go]: done

File: issue.go (hunk 9 above)
Comment (sridhar02) [reply issue.go:232]: done

**Pull request merged**